### PR TITLE
Implemented LRTA* search Agent

### PR DESCRIPTION
**Pull request by SnShine**
Changes:
* Implementes LRTA* agent
* Adds skeleton problem classes for Maze [4.19] problem and 1-dim State Space problem [4.23].

@norvig, is it okay to have these new classes for problems? If it is alright, I will go on implementing these subproblem classes.

### CODE REVIEW

File: search.py [hunk 1]
```diff
@@ -456,12 +471,77 @@ def __call__(self, percept):
...
+
+    def __init__(self, problem):
+        self.problem = problem
+        self.result = {}
+        self.H = {}
+        self.s = None
+        self.a = None
+
+    def __call__(self, s1):
+        current_state = self.update_state(s1)
```
Comment (reachtarunhere) [inline search.py]: @SnShine I guess there is a slight error here. In your implementation there is no update_state method present but it is being called.

**Comment by reachtarunhere** [comment]
@SnShine I think we should try to keep the code uniform. By that I mean that my implementation of OnlineDFS agent is using percept and your implementation is using s1. We can obviously decide which one to keep and then either of us can make changes.

File: search.py [hunk 2]
```diff
@@ -456,12 +471,77 @@ def __call__(self, percept):
...
+        if self.problem.goal_test(current_state):
+            self.a = None
+        else:
+            if current_state not in H:
+                H[current_state] = h(current_state)
```
Comment (reachtarunhere) [inline search.py]: I am not sure about this but I couldn't find h() defined anywhere.

File: search.py [hunk 3]
```diff
@@ -456,12 +471,77 @@ def __call__(self, percept):
...
+            s = current_state
+            return a
+
+    def LRTA_cost(s, a, s1, H):
+        """
+        returns cost to move from state 's' to state 's1' plus
+        estimated cost to get to goal from s1
+        """
+        if s1 is None:
+            return(h(s))
```
Comment (reachtarunhere) [inline search.py]: @SnShine Same as above pardon me if I am wrong with this where is h() defined?

**Comment by SnShine** [comment]
@reachtarunhere I am following the pseudo-code from the book where 's1' is used. I could've gone with percept but there is also 's', previous state. So, I stick with the names used in pseudo-code from 3rd edition.

**Comment by SnShine** [comment]
I am a bit confused if we need the new problem subclasses, MazeProblem and StateSpace problem. If we decide not to have them, I will put unimplemented methods (which we need to override) update_state(), c(), h() in LRTA Agent.

**Comment by SnShine** [comment]
If we are having those problem subclasses, h() becomes self.problem.h() and c(), update_state() accordingly.

**Comment by reachtarunhere** [comment]
@SnShine I agree your code is more closely following the pseudo-code. Still I have a personal preference for percept. Once this PR gets merged I will change my code to look like yours.

**Pull request closed without merging**